File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        lo=0
        hi=len(nums)-1

        while lo<=hi:
            mid=lo+(hi-lo)//2
            if nums[mid]==target:
                return True
            elif nums[mid]==nums[lo]==nums[hi]:
                lo=lo+1
                hi=hi-1
            elif nums[mid]>=nums[lo]:
                if target>=nums[lo] and target<nums[mid]:
                    hi=mid-1
                else:
                    lo=mid+1
            elif nums[mid]<=nums[hi]:
                if target>nums[mid] and target<=nums[hi]:
                    lo=mid+1
                else:
                    hi=mid-1
        return False
        

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {boolean}
 */
var search = function(nums, target) {
    let lo=0
    let hi=nums.length-1
    let ans=false
    while(lo<=hi){
        let mid=Math.floor((lo+hi)/2)
        if(nums[mid]===target){
            return true;
        }else if (nums[lo]===nums[mid] && nums[mid]===nums[hi]){
            lo=lo+1
            hi=hi-1
        }else if (nums[lo]<=nums[mid]){
            if(target >=nums[lo] && target <nums[mid]){
                hi=mid-1
            }else{
                lo=mid+1
            }
        }else{
            if(target>= nums[mid] && target<=nums[hi]){
                lo=mid+1
            }else{
                hi=mid-1
            }
        }
    }
    return ans
    
};